compiler/damlc/daml-compiler/src/DA/Daml/Compiler/DataDependencies.hs

```diff
@@ -467,48 +523,53 @@ isConstraint = \case
 genModule :: Env -> LF.PackageRef -> LF.ModuleName -> Gen Module
 genModule env pkgRef modName = do
     let Config{..} = envConfig env
-        isStable
-            | LF.PRSelf <- pkgRef = False
-            | LF.PRImport pkgId <- pkgRef =
-                pkgId `Set.member` configStablePackages
+        (packageQualified, modName') = case pkgRef of
+            LF.PRImport pkgId
+                | pkgId `Set.member` configDependencyPackages -> (False, prefixDependencyModule pkgId modName)
+                | pkgId `Set.member` configStablePackages -> (False, prefixModuleName configSdkPrefix modName)
+            _ -> (True, modName)
         unitId = configGetUnitId pkgRef
-    genModuleAux env isStable unitId modName
+    genModuleAux packageQualified unitId modName'
 
 genStableModule :: Env -> UnitId -> LF.ModuleName -> Gen Module
-genStableModule env = genModuleAux env True
-
-genModuleAux :: Env -> Bool -> UnitId -> LF.ModuleName -> Gen Module
-genModuleAux env isStable unitId modName = do
-    let sdkPrefix = configSdkPrefix (envConfig env)
-        newModName
-            | isStable = LF.ModuleName (sdkPrefix ++ LF.unModuleName modName)
-            | otherwise = modName
-        ghcModName = mkModuleName . T.unpack $ LF.moduleNameString newModName
-        modRef = ModRef isStable unitId newModName
+genStableModule env unitId = genModuleAux False unitId . prefixModuleName (configSdkPrefix $ envConfig env)
+
+prefixModuleName :: [T.Text] -> LF.ModuleName -> LF.ModuleName
+prefixModuleName prefix (LF.ModuleName mod) = LF.ModuleName (prefix <> mod)
+
+prefixDependencyModule :: LF.PackageId -> LF.ModuleName -> LF.ModuleName
+prefixDependencyModule (LF.PackageId pkgId) = prefixModuleName ["Pkg_" <> pkgId]
+
+genModuleAux :: Bool -> UnitId -> LF.ModuleName -> Gen Module
+genModuleAux isQualified unitId modName = do
+    let ghcModName = mkModuleName . T.unpack $ LF.moduleNameString modName
+        modRef = ModRef isQualified unitId modName
     emitModRef modRef
     pure $ mkModule unitId ghcModName
 
-convType :: Env -> LF.Type -> Gen (HsType GhcPs)
-convType env =
+convType :: Env -> MS.Map LF.TypeSynName LF.PackageId -> LF.Type -> Gen (HsType GhcPs)
+convType env reexported =
```

Comment:
Happy leave it for a future PR, but I think it makes sense to pass `reexported` inside `env` at this point.

Reply:
good idea, I’ll leave it for a separate PR.